crates/core/src/main.rs

```diff
@@ -29,5 +34,44 @@ pub extern "C" fn init() {
 fn main() {
     let bytecode = unsafe { BYTECODE.take().unwrap() };
     let runtime = unsafe { RUNTIME.take().unwrap() };
-    execution::run_bytecode(&runtime, &bytecode)
+    execution::run_bytecode(&runtime, &bytecode);
+}
+
+// Removed in post_processing.
+/// 1. Allocate memory of new_size with alignment.
+/// 2. If original_ptr != 0
+///   a. copy min(new_size, original_size) bytes from original_ptr to new memory
+///   b. de-allocate original_ptr
+/// 3. return new memory ptr
+///
+/// # Safety
+///
+/// * `original_ptr` must be 0 or a valid pointer
+/// * if `original_ptr` is not 0, it must be valid for reads of `original_size`
+///   bytes
+/// * if `original_ptr` is not 0, it must be properly aligned
+/// * if `original_size` is not 0, it must match the `new_size` value provided
+///   in the original `canonical_abi_realloc` call that returned `original_ptr`
+#[export_name = "canonical_abi_realloc"]
+pub unsafe extern "C" fn canonical_abi_realloc(
+    original_ptr: *mut u8,
+    original_size: usize,
+    alignment: usize,
+    new_size: usize,
+) -> *mut std::ffi::c_void {
+    alloc::canonical_abi_realloc(original_ptr, original_size, alignment, new_size)
+}
+
+// Removed in post-processing.
+/// Evaluates QuickJS bytecode and invokes the exported JS function name.
+///
+/// # Safety
+///
+/// * `fn_name_ptr` must reference a UTF-8 string with `fn_name_size` byte
+///   length.
+#[export_name = "javy.invoke"]
+pub unsafe extern "C" fn invoke(fn_name_ptr: *mut u8, fn_name_size: usize) {
+    let js_fn_name = str::from_utf8_unchecked(slice::from_raw_parts(fn_name_ptr, fn_name_size));
+    let runtime = unsafe { RUNTIME.take().unwrap() };
+    execution::invoke_function(&runtime, FUNCTION_MODULE_NAME, js_fn_name);
```

Comment:
Interestingly enough, if I run `execution::run_bytecode(runtime, bytecode);` in here before `invoke_function`, I see the top-level logic getting executed twice in the tests. However, it works as expected in `lib.rs` when `run_bytecode` is executed followed by `invoke_function`.